src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,177 @@
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix,
+                                async ctx => await FixCodeTwoParameters(context.Document, root, node, ctx).ConfigureAwait(false),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix),
+                            diagnostic);
+                }
+            }
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var typedNodeToFix = (InvocationExpressionSyntax)nodeToFix;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            SyntaxNode stringArgument = generator.IdentifierName(fixComponents.StringArgumentName);
+            SyntaxNode lengthNode = generator
+                .MemberAccessExpression(
+                    stringArgument, 
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode, 
+                    fixComponents.OriginalInnerArguments[0].Expression);
+
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                stringArgument,
+                fixComponents.OriginalInnerArguments[0], 
+                lengthArgument);
+            var newRoot = root
+                .ReplaceNode(
+                    typedNodeToFix, 
+                    newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private struct FixComponents
+        {
+            public FixComponents(
+                string stringArgumentName, 
+                SyntaxNode targetMethod, 
+                SeparatedSyntaxList<ArgumentSyntax> originalInnerArguments)
+            {
+                StringArgumentName = stringArgumentName;
+                TargetMethod = targetMethod;
+                OriginalInnerArguments = originalInnerArguments;
+            }
+
+            public string StringArgumentName { get; }
+            public SyntaxNode TargetMethod { get; }
+            public SeparatedSyntaxList<ArgumentSyntax> OriginalInnerArguments { get; }
+        }
+
+        private static async Task<Document> FixCodeTwoParameters(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var typedNodeToFix = (InvocationExpressionSyntax) nodeToFix;
+            var generator = SyntaxGenerator.GetGenerator(document);
```

Comment:
instead of casting directly to a C# type, just use semanticModel.GetOperation(nodeToFix) and work with the language agnostic IOperatoin type you get back.